src/hub/vendor/abstract.py

```diff
@@ -47,3 +53,20 @@ def run(self) -> None:
 
     def create_data(self, **kwargs) -> Dict[str, str]:
         return dict(event_id=self.payload.id, event_type=self.payload.type, **kwargs)
+
+    def customer_event_to_all_routes(self, data_projection, data):
+        subsets = []
+        for route in data_projection:
+            try:
+                logger.debug("sending to", key=route)
+                subset = dict((k, data[k]) for k in data_projection[route] if k in data)
+                payload = {"type": route, "data": json.dumps(subset)}
+                subsets.append(payload)
+                logger.info("subset", subset=subset)
+
+                logger.debug("sent to", key=route)
+            except Exception as e:
+                # log something and maybe change the exception type.
+                logger.error("projection exception", error=e)
+            finally:
+                self.send_to_all_routes(subsets)
```

Comment:
Reading through this, it looks like this call should be outside of the for loop since it is defined above the logic. It appears that through each iteration data is appended to subsets, so it seems like the idea is for all of the subset data should be sent at once rather than at the end of each iteration. Is that the case?

Reply:
So the point of this function is to gather all the subsets and send to all routes, if none, 1 or more than 1 are gathered then the send_to_all_routes is called at the end of the try/except.  If nothing is added in subsets then it's sent with an empty list and no work will be done by the pipeline.

Reply:
```
def customer_event_to_all_routes(self, data_projection, data):
        subsets = []
        for route in data_projection:
            try:
                logger.debug("sending to", key=route)
                subset = dict((k, data[k]) for k in data_projection[route] if k in data)
                payload = {"type": route, "data": json.dumps(subset)}
                subsets.append(payload)
                logger.info("subset", subset=subset)

                logger.debug("sent to", key=route)
            except Exception as e:
                # log something and maybe change the exception type.
                logger.error("projection exception", error=e)
        self.send_to_all_routes(subsets)
```

Reply:
I see what you mean now.  Will update.

Reply:
@jackiemunroe I moved the send_to_all_routes to the else on the for loop, there it will be called after the for loop completes.